### ISSUE TITLE
what is the best practice to integrate orml?

### ISSUE THREAD

**Issue by 2075** [open-web3-stack/open-runtime-module-library#380]
hey there, thanks for the great work so far. i am wondering, what would the best way to integrate orml, as we are having a hard time it seems to solve this.

our testnet is still on 2.0.1 while the local development uses >3.0.0 with local pallets, frame, etc.

therefore it usually happens, when using orml from git, the versions are unpar. when using a local mirror, we can change the paths to use our local ones, but then the diff to substrate and orml versions consumes more time. with 2.0.1 it was also unresolvable to get matching versions (compiling) with our local substrate and orml, whether through github or local submodule.

how do you guys solve it, or are you all using substrate bleeding edge?

thanks a lot for input,
marco from zero.io

**Comment by xlc** [comment]
You need to pick a version.

If you are using Substrate 2.0.1, use 0.3.0 release from crates.io.
If you are using Substrate 3.0.0, use 0.4.0 release from crates.io.

If you are using Substrate master / rococo-v1, use orml master / rococo-v1.

You can checkout projects using ORML and see their configuration https://github.com/open-web3-stack/open-runtime-module-library#projects-using-orml

**Comment by 2075** [comment]
hey @xlc 

thanks for your feedback. we already understood that roughly 0.3.0 should work with 2.0.1 and 0.4.0 with 3.0.0.
sadly we have changes in our core and using the full on substrate as a base didn't make it easier to port changes across the versions of substrate.

we also checked the mentioned projects, some great code out there, but really not feasible to back-/port all the time ( especially when we even cannot get a pinned version setup running :) )

even when we build e.g. a modified complete substrate node using git for resolving the packages isntead of the local ones - without our changes - and using e.g. orml from crates.io with respective versions results in a bazillion of version conflicts. after giving this a shot since roughly 0.3.0 and following, we gave up for now.

thanks anyway and keep up the inspiring work.

**Comment by xlc** [comment]
You need to either all use creates.io version or all use Github version.

In Acala our Cargo.toml have patches to ensure consist version of everything.

Without seeing your code I can't give more advice.

**Comment by 2075** [comment]
yes, I saw that, for that reason we started to work with a local submodule orml and modified the paths to use our local versions. in the current branch, we are using the pinned versions from acala master and orml from crates, the last missing bit seems to be contracts pallet which creates a version mismatch with parity-util-mem, but just now i saw that the 3.0.0 seems to be on the way into master. 

i am in the lucky situation to always try to sync the versions, whenever a new version seems to be finalised in substrate master :D 

regarding code: a mirror is available on github containing various attempts of syncing the versions in our codebase, dating back to when the 2.0.0 was released. i would not recomment to check that, though, as it will just make your compiler unhappy..

thanks again for your help, I will ping back as soon as there is better information available and i can ask better questions.
